### PR TITLE
Rework how comments are represented in the Piece model

### PR DESCRIPTION
**Pull request by munificent**
I've been frustrated with how the new formatter represents comments for a while. Jamming them inside the `_lines` list in TextPiece sort of works, but it means we lose track of what's a comment and what isn't. It means PieceWriter and TextPiece are more mutable than I'd like, which is I think a large part of why the logic of those classes is hard to follow.

I was trying to fix some bugs that I noticed when migrating the regression tests. For example:

```dart
value =
    // comment
    a + b;
```

Currently, the formatter outputs:

```dart
value =
    // comment
    a +
        b;
```

That's because the `// comment` is part of the TextPiece for `a`, so it thinks there's a newline inside the first operand to `+` and thus the `+` splits.

Trying to fix that with the current comment handling was really hard. So I finally dug in and tried to come up with a cleaner way to model comments. This PR does that. The basic idea is that we make each comment a separate piece. That in turn implies that TextPieces for non-comment code needs to support child pieces for the comments attached to them.

It doesn't, like, radically improve comment handling. (It also doesn't fix the above bug yet.) But I do think the resulting code is cleaner, and I have a separate fix for the above bug that builds on it that I think would have been harder to do without.

Also, I have a goal to do an optimization where we coalesce adjacent TextPiece and SpacePieces when we can to produce a smaller, chunkier piece tree. In my informal tests, it measurably improves perf. But I've never been able to get a fully working implementation because there are so many corner cases around the mutability in PieceWriter that it was a bug farm. I hope this PR will make that more feasible.

### CODE REVIEW

File: test/tall/statement/for_comment.stmt [hunk 1]
```diff
@@ -69,8 +69,7 @@ for  (  init
 <<<
 ### Looks weird, but users don't tend to put comments here.
 for (
-  init
-  // comment
+  init// comment
```
Comment (kallentu) [inline test/tall/statement/for_comment.stmt:72]: Do we want a little bit of spacing (1 space) when we move a comment like this?

**Review by kallentu — APPROVED** [review]

File: lib/src/front_end/piece_writer.dart [hunk 2]
```diff
@@ -74,119 +80,84 @@ class PieceWriter {
...
-
-    // Otherwise, there are no comments, or all comments are hanging off the
-    // previous TextPiece.
-    return null;
-  }
+  /// If [multiline] is `true`, then [token]'s lexeme may contain internal
+  /// newlines. The lexeme will be split into separate lines. If omitted, then
+  /// [token] must not contain newlines.
+  ///
+  /// If [lexeme] is given, then uses that lexeme string instead of [token]'s
```
[the remark this answers is not in the document]
Comment (munificent) [reply lib/src/front_end/piece_writer.dart:107]: Done.

File: lib/src/front_end/piece_writer.dart [hunk 3]
```diff
@@ -74,119 +80,84 @@ class PieceWriter {
...
-  }
-
-  /// Writes [text] to the current [TextPiece].
-  ///
-  /// If [offset] is given and it contains any selection markers, then attaches
-  /// those markers to the [TextPiece].
-  void _write(String text, int offset, {bool multiline = false}) {
-    _currentText.append(text,
+  /// Appends [text] to [piece] and updates any selection markers that fall
+  /// within it.
```
[the remark this answers is not in the document]
Comment (munificent) [reply lib/src/front_end/piece_writer.dart:161]: Could idea. Done.

File: test/tall/statement/for_comment.stmt (hunk 1 above)
Comment (munificent) [reply test/tall/statement/for_comment.stmt]: Ideally, it would be a newline as in the previous code. I tried a bunch last night and couldn't figure out a clean way to get that working. But I gave it another shot today and found a solution, so done. :)

Thanks for pushing on this.
